### ISSUE TITLE
GRPO trainer is not compatible with FSDP

### ISSUE THREAD

**Issue by Tavish9** [huggingface/trl#3348]
### Reproduction

```python
from trl import GRPOConfig, GRPOTrainer, TrlParser

model_name_or_path = "Qwen/Qwen2.5-0.5B-Instruct"
parser = TrlParser(GRPOConfig)
training_args, = parser.parse_args_into_dataclasses()

trainer = GRPOTrainer(model=model_name_or_path, reward_funcs=[], args=training_args)

print(trainer.accelerator.state.device)
print(trainer.model.device)
print(trainer.ref_model.device)

```

`torchrun --nnodes=1 --nproc_per_node=1 test.py` will output 
```
cuda:0
cuda:0
cuda:0
```

`torchrun --nnodes=1 --nproc_per_node=1 test.py --fsdp "full_shard"`  will output 
```
cuda:0
cpu
cpu
```

Then, in Transformers.trainer._inner_training_loop, accelerator would put `self.model` to correct device. However, `self.ref_model` is left since it's defined in GRPOTrainer.

During training, we got
```
self.accelerator.state.device -> cuda
self.model.device -> cuda
self.ref_model.device -> cpu
```

### System Info

- Platform: Linux-5.10.134-008.12.kangaroo.al8.x86_64-x86_64-with-glibc2.35
- Python version: 3.10.17
- TRL version: 0.17.0.dev0
- PyTorch version: 2.6.0
- CUDA device(s): NVIDIA A100-SXM4-80GB
- Transformers version: 4.51.2
- Accelerate version: 1.6.0
- Accelerate config: not found
- Datasets version: 3.5.0
- HF Hub version: 0.30.2
- bitsandbytes version: not installed
- DeepSpeed version: 0.16.7
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.75.0
- PEFT version: 0.15.2
- vLLM version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by Tavish9** [comment]
I don’t know the proper way, either. But i know how to fix it in my local env.

I would like to help if you or the author of trl can share the design style with me🤗

**Comment by shirinyamani** [comment]
#3354

**Comment by sorgfresser** [comment]
#3354 does not address this so far, the same issue persists on the checked out branch.

**Comment by sorgfresser** [comment]
The issue mentioned here relates to the ref_model on CPU, which is not addressed by gathering. It stems from the evaluation mode passed to `prepare_model`. Using
```
                self.ref_model = self.accelerator.prepare(self.ref_model)
                self.ref_model.eval()
```
instead fixes this and resolves the issue, though it will introduce more memory consumption (as the model is now on the GPU rather than the CPU).

**Comment by shirinyamani** [comment]
@sorgfresser Oh I see, Thanks for enlightening me! 
So the issue is the policy model is moved to the cuda device, however, the ref_model is on CPU, is this correct?

**Comment by sorgfresser** [comment]
Yep! :)

**Comment by LeonEricsson** [comment]
> [#3354](https://github.com/huggingface/trl/pull/3354) does not address this so far, the same issue persists on the checked out branch.

Are you sure? My ref_model seems to be correctly assigned inside `_generate_and_score_completions`:


```
model device: cuda:0
ref model device: cuda:0
```

**Comment by sorgfresser** [comment]
Could you share the code you used for this? If I launch via accelerate with FSDP version 1, I get a ref model on CPU.

**Comment by LeonEricsson** [comment]
> Could you share the code you used for this? If I launch via accelerate with FSDP version 1, I get a ref model on CPU.

Interesting, I am indeed experiencing the same problem using `accelerate launch`. Using

`torchrun --nnodes=1 --nproc_per_node=1 grpo.py --fsdp "full_shard"`

works for me.

**Comment by LeonEricsson** [comment]
> Could you share the code you used for this? If I launch via accelerate with FSDP version 1, I get a ref model on CPU.

Apparently there is an earlier PR that already added FSDP support #3260. That branch correctly moves `ref_model`.
